### PR TITLE
Phoenix 5976 Cannot drop a column when the index view is involved

### PR DESCRIPTION
**Pull request by yanxinyi**

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java [hunk 1]
```diff
@@ -411,5 +412,126 @@ public void testDroppingIndexedColDropsIndex() throws Exception {
...
+                stmt.execute(
+                        "CREATE VIEW IF NOT EXISTS " + view + "  (PK1 DATE NOT NULL,PK2 CHAR(15) NOT NULL," +
+                                "NON_PK1 CHAR(15),NON_PK2 CHAR(15) CONSTRAINT PKVIEW PRIMARY KEY (PK1,PK2)) " +
+                                "AS SELECT * FROM " + table + " WHERE KEY_PREFIX = '123'");
+                // create index
+                stmt.execute("CREATE INDEX " + index + " ON " + view + " (PK2, PK1) INCLUDE (NON_PK1, NON_PK2)");
+                // drop column
+                stmt.execute("ALTER VIEW "  + view + " DROP COLUMN NON_PK1");
+            }
+        } catch (Exception e) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java:437]: No need to catch here

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java [hunk 2]
```diff
@@ -411,5 +412,126 @@ public void testDroppingIndexedColDropsIndex() throws Exception {
...
+                stmt.execute(
+                        "CREATE VIEW IF NOT EXISTS "  + view + "  (PK1 DATE NOT NULL,PK2 CHAR(15) NOT NULL," +
+                                "NON_PK1 CHAR(15),NON_PK2 CHAR(15) CONSTRAINT PKVIEW PRIMARY KEY (PK1,PK2)) " +
+                                "AS SELECT * FROM "  + table + " WHERE KEY_PREFIX = '123'");
+                // create index
+                stmt.execute("CREATE INDEX " + index + " ON " + view + " (PK2, PK1) INCLUDE (NON_PK1, NON_PK2)");
+                // drop column
+                stmt.execute("ALTER VIEW "  + view + " DROP COLUMN NON_PK1");
+            }
+        } catch (Exception e) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java:463]: Same here, no need to catch exception and fail. Just let it propagate up

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java [hunk 3]
```diff
@@ -411,5 +412,126 @@ public void testDroppingIndexedColDropsIndex() throws Exception {
...
+            conn.setAutoCommit(true);
+            try (Statement stmt = conn.createStatement()) {
+                stmt.execute("CREATE TABLE IF NOT EXISTS " + table + " (ID CHAR(10) NOT NULL,KEY_PREFIX CHAR(3) NOT NULL, NON_PK1 CHAR(15),NON_PK2 CHAR(15), " +
+                        " CONSTRAINT PK PRIMARY KEY (ID,KEY_PREFIX))");
+                // create index
+                stmt.execute("CREATE INDEX " + index + " ON " + table +" (KEY_PREFIX, ID) INCLUDE (NON_PK1, NON_PK2)");
+                // drop column
+                stmt.execute("ALTER TABLE " + table + " DROP COLUMN NON_PK1");
+            }
+        } catch (Exception e) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java:483]: ditto

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java [hunk 4]
```diff
@@ -411,5 +412,126 @@ public void testDroppingIndexedColDropsIndex() throws Exception {
...
+                        " CONSTRAINT PK PRIMARY KEY (ID,KEY_PREFIX)) MULTI_TENANT=TRUE");
+                // create view
+                stmt.execute(
+                        "CREATE VIEW IF NOT EXISTS " +  view  + "  (PK1 DATE NOT NULL,PK2 CHAR(15) NOT NULL," +
+                                "NON_PK1 CHAR(15),NON_PK2 CHAR(15) CONSTRAINT PKVIEW PRIMARY KEY (PK1,PK2)) " +
+                                "AS SELECT * FROM " + table + " WHERE KEY_PREFIX = '123'");
+                // drop column
+                stmt.execute("ALTER VIEW " + view + " DROP COLUMN NON_PK3");
+            }
+        } catch (Exception e) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java:506]: ditto

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java [hunk 5]
```diff
@@ -411,5 +412,126 @@ public void testDroppingIndexedColDropsIndex() throws Exception {
...
+                        "CREATE VIEW IF NOT EXISTS " +  view  + "  (PK1 DATE NOT NULL,PK2 CHAR(15) NOT NULL," +
+                                "NON_PK1 CHAR(15),NON_PK2 CHAR(15) CONSTRAINT PKVIEW PRIMARY KEY (PK1,PK2)) " +
+                                "AS SELECT * FROM " + table + " WHERE KEY_PREFIX = '123'");
+
+                stmt.execute("CREATE INDEX " + index + " ON " + view +" (PK2, PK1) INCLUDE (NON_PK1, NON_PK2)");
+
+                // drop column
+                stmt.execute("ALTER VIEW " + view + " DROP COLUMN NON_PK3");
+            }
+        } catch (Exception e) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/index/DropColumnIT.java:533]: ditto

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java [hunk 6]
```diff
@@ -80,7 +81,8 @@ public DropColumnMutator(Configuration conf) {
     }
 
     @Override
-    public MutateColumnType getMutateColumnType() {
+    public MutateColumnType getMutateColumnType()
+    {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java:85]: nit: remove this diff

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java [hunk 7]
```diff
@@ -181,13 +183,28 @@ public MetaDataMutationResult validateAndAddMetadata(PTable table,
...
+                        if (Bytes.compareTo(indexSchema, rowKeyMetaData[SCHEMA_NAME_INDEX]) == 0
+                                && Bytes.compareTo(indexTableName, rowKeyMetaData[TABLE_NAME_INDEX]) == 0) {
+                            column = MetaDataUtil.getColumn(pkCount, rowKeyMetaData, indexTable);
+                            break;
+                        }
+                    }
+                }
                 if (column == null)
                     continue;
                 // ignore any puts that modify the ordinal positions of columns
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java:205]: Do we know why we are ignoring puts that modify ordinal positions?
FYI @twdsilva

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java [hunk 8]
```diff
@@ -181,13 +183,28 @@ public MetaDataMutationResult validateAndAddMetadata(PTable table,
...
             int pkCount = getVarChars(key, rowKeyMetaData);
             if (isView && mutation instanceof Put) {
-                PColumn column = MetaDataUtil.getColumn(pkCount, rowKeyMetaData, table);
+                PColumn column = null;
+                // checking put from the view or index
+                if (Bytes.compareTo(schemaName, rowKeyMetaData[SCHEMA_NAME_INDEX]) == 0
+                        && Bytes.compareTo(tableName, rowKeyMetaData[TABLE_NAME_INDEX]) == 0) {
+                    column = MetaDataUtil.getColumn(pkCount, rowKeyMetaData, table);
+                } else {
+                    for(int i = 0; i < table.getIndexes().size(); i++) {
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java:192]: Is this required even when the parent is a table (table_type='u') or only needed when it is a view (table_type='v')?

Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java:192]: Synced offline with @yanxinyi . Missed that we have an `isView` check in the outer if condition. Thanks

**Comment by yanxinyi** [comment]
Added two new tests
1. drop a column from the table and this column is included in the index table.
2. drop a column from the table with a child view.


@ChinmaySKulkarni as we discuss offline, I'm going to create another JIRA for fixing the ordinal position issue.

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java (hunk 7 above)
Comment (twdsilva) [reply phoenix-core/src/main/java/org/apache/phoenix/coprocessor/DropColumnMutator.java:204]: I think this was to handle older clients that update ordinal column positions when a column is dropped.

**Pull request closed without merging**